### ISSUE TITLE
CA1506 False positive

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn-analyzers#2970]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/704043/ca1506-false-positive.html)._

---
This thing is way off, it doesn't seem to understan interfaces or inheritance and is confused.
This class is obviosly not coupled with 36 different types.

See the references window (Yes I know it's broke too &quot;By design&quot;)

Severity	Code	Description	Project	File	Line	Suppression State
Warning	CA1506	'SAPChangeProcessorShipTo.ProcessFile(string, IJediRepositoryFactory)' is coupled with 36 different types from 15 different namespaces. Rewrite or refactor the method to decrease its class coupling, or consider moving the method to one of the other types it is tightly coupled with. A class coupling above 40 indicates poor maintainability, a class coupling between 40 and 30 indicates moderate maintainability, and a class coupling below 30 indicates good maintainability.	KerryConnect	C:\git\jEDI\KerryConnect\SAP\SAPChangeProcessorShipTo.cs	20	Active

---
#### Visual Studio Feedback System on 8/22/2019, 03:39 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Visual Studio Feedback System on 8/22/2019, 10:43 AM: 

<p>This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved.</p>

---
### Original Solutions
(no solutions)

**Comment by wizofaus** [comment]
I've see this a lot on, e.g. ASP.NET controller classes that have maybe 5 injected dependencies and where each function simply calls a method on said dependencies.  I don't know what it's counting - .NET core framework classes??

**Comment by Evangelink** [comment]
Relates to #2133. @mavasani I think this can be closed as it was solved through the other ticket.

**Comment by Evangelink** [comment]
@mavasani I don't know if that's the only issue, I am having a lot of these warnings when doing MVVM applications. When I have looked at the actual dependencies it seems to be looking recursively and I am not entirely sure if that's what we want or not. For example if on the class C you reference class A which references let's say 50 types, C is considered as referencing all of A's references + A.

**Comment by wizofaus** [comment]
One example I came across today, we have CA1825 turned on - avoid zero-length array allocations, with the recommendation to use Array.Empty<type>() instead.
But when I changed

```var foo = new Guid[];```

to

```var foo = Array.Empty<Guid>();```

CA1506 fired.  How does using the .NET Array.Empty( ) increase my coupling at all?

**Comment by mavasani** [comment]
@wizofaus I believe this has been addressed by https://github.com/dotnet/roslyn-analyzers/pull/3233, which is available in latest 3.3.0 beta package. https://github.com/dotnet/roslyn-analyzers/blob/0a6e4b3935bbcdde79eb90d1760368e89ffff826/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/CodeMetricsAnalyzer.cs#L151-L158

@Evangelink Can you confirm if this issue should now be closed?

**Comment by Evangelink** [comment]
@mavasani I'll make a test but I don't think it's done, on your reference we did it for `CA1501` here it is about `CA1506`.

**Comment by mavasani** [comment]
> on your reference we did it for CA1501 here it is about CA1506.

Ah, thanks. I will let you drive this issue forward!

**Comment by Evangelink** [comment]
> One example I came across today, we have CA1825 turned on - avoid zero-length array allocations, with the recommendation to use Array.Empty() instead.
> But when I changed
> 
> `var foo = new Guid[0];`
> 
> to
> 
> `var foo = Array.Empty<Guid>();`
> 
> CA1506 fired. How does using the .NET Array.Empty( ) increase my coupling at all?

@wizofaus Debugging the rule, I can see that the main difference between `Guid[0]` and `Array.Empty<Guid>()` is that in the first case the array part is ignored. I don't think it shall. @mavasani Are you ok to count `[]` as a reference to `System.Array`?

**Comment by Evangelink** [comment]
Other question, do we want to count 2 coupling for `double` and `double?` or only 1?

**Comment by wizofaus** [comment]
Why is referring to `double` a coupling at all? It's part of the language?  And System.Array is part of the core .NET API (System.Runtime.dll), again I wouldn't expect it to count as a coupling at all.

**Comment by Evangelink** [comment]
@wizofaus Just checked the code and primitive types don't count but `DateTime` or `TimeSpan` do.

Getting back to my question, do we agree that in the following example the coupling shall be 1?

```csharp
public class C // Coupling = 1 (DateTime)? or Coupling = 2 (DateTime + Nullable<T>)?
{
	private DateTime d1;
	private DateTime? d2;
}
```

Regarding the array, I am a bit in 2 state of mind. Somehow I think array should not count but then why do we count `List<T>` or `ImmutableArray<T>`...?

**Comment by mavasani** [comment]
Going to close this issue, please re-open if we feel any more work is needed here.
